mlflow/server/js/src/components/ParallelCoordinatesPlotView.js

```diff
@@ -0,0 +1,188 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import Plot from 'react-plotly.js';
+import PropTypes from 'prop-types';
+import Utils from '../utils/Utils';
+import _ from 'lodash';
+
+const AXIS_LABEL_CLS = '.pcp-plot .parcoords .y-axis .axis-heading .axis-title';
+
+export class ParallelCoordinatesPlotView extends React.Component {
+  static propTypes = {
+    runUuids: PropTypes.arrayOf(String).isRequired,
+    paramKeys: PropTypes.arrayOf(String).isRequired,
+    metricKeys: PropTypes.arrayOf(String).isRequired,
+    paramDimensions: PropTypes.arrayOf(Object).isRequired,
+    metricDimensions: PropTypes.arrayOf(Object).isRequired,
+  };
+
+  state = {
+    // Current sequence of all axes, both parameters and metrics.
+    sequence: [...this.props.paramKeys, ...this.props.metricKeys],
+  };
+
+  static getDerivedStateFromProps(props, state) {
+    const keysFromProps = [...props.paramKeys, ...props.metricKeys];
+    const keysFromState = state.sequence;
+    if (!_.isEqual(_.sortBy(keysFromProps), _.sortBy(keysFromState))) {
+      return { sequence: keysFromProps };
+    }
+    return null;
+  }
+
+  getData() {
+    const { sequence } = this.state;
+    const { paramDimensions, metricDimensions, metricKeys } = this.props;
+    const lastMetricKey = this.findLastKeyFromState(metricKeys);
+    const lastMetricDimension = this.props.metricDimensions.find((d) => d.label === lastMetricKey);
+    const colorScaleConfigs = ParallelCoordinatesPlotView.getColorScaleConfigsForDimension(
+      lastMetricDimension,
+    );
+    // This make sure axis order consistency across renders.
+    const orderedDimensions = ParallelCoordinatesPlotView.getDimensionsOrderedBySequence(
+      [...paramDimensions, ...metricDimensions],
+      sequence,
+    );
+    return [
+      {
+        type: 'parcoords',
+        line: { ...colorScaleConfigs },
+        dimensions: orderedDimensions,
+      },
+    ];
+  }
+
+  static getDimensionsOrderedBySequence(dimensions, sequence) {
+    return _.sortBy(dimensions, [(dimension) => sequence.indexOf(dimension.label)]);
+  }
+
+  static getLabelElementsFromDom = () => Array.from(document.querySelectorAll(AXIS_LABEL_CLS));
+
+  static getSequenceFromDom = () =>
+    ParallelCoordinatesPlotView.getLabelElementsFromDom().map((el) => el.innerHTML);
+
+  findLastKeyFromState(keys) {
+    const { sequence } = this.state;
+    const keySet = new Set(keys);
+    return _.findLast(sequence, (key) => keySet.has(key));
+  }
+
+  findLastMetricFromDom() {
+    const sequence = ParallelCoordinatesPlotView.getSequenceFromDom();
+    const metricsKeySet = new Set(this.props.metricKeys);
+    return _.findLast(sequence, (key) => metricsKeySet.has(key));
+  }
+
+  static getColorScaleConfigsForDimension(dimension) {
```

Comment:
nit: `getColorScaleConfigForDimension`

Reply:
I'm not sure about this because this function actually returns multiple configs in but in one object.